schedulefree/adamw_schedulefree.py

```diff
@@ -124,39 +129,67 @@ def step(self, closure=None):
             if not group['train_mode']:
                 raise Exception("Not in train mode!")
 
-            for p in group['params']:
-                if p.grad is None:
-                    continue
+            if group['foreach']:
+                for p in group['params']:
+                    if 'z' not in self.state[p] and p.grad is not None:
+                        self.state[p]['z'] = torch.clone(p.data)
+                        self.state[p]['exp_avg_sq'] = torch.zeros_like(p.data)
+
+                y, grad, exp_avg_sq, z = zip(*[(p.data, p.grad, self.state[p]['exp_avg_sq'], self.state[p]['z']) 
+                                               for p in group['params'] if p.grad is not None])
+    
+                torch._foreach_mul_(exp_avg_sq, beta2)
+                torch._foreach_addcmul_(exp_avg_sq, grad, grad, value=1-beta2)
+                denom = torch._foreach_add_(torch._foreach_sqrt(exp_avg_sq), eps)
```

Comment:
Errors for me, `torch._foreach_add_` is inplace operation, should it be `_foreach_add` ? And inplace operation is also used at the line 

`grad_normalized = torch._foreach_div_(grad, denom)`

Reply:
Very strange that that errors, I have a training run active right now that is running on this code (except in a crudely hardcoded version rather than a properly implemented option).  I used inplace operations wherever inplace operations were used in the original code, and this corresponds to this line, and I think that that line is a proper conversion:
`denom = exp_avg_sq.sqrt().add_(eps)`

All of the operations beyond that line are inplace and are hardly ones that I would suspect would misbehave since they are very simple cases.  They definitely should remain inplace, too, for the sake of saving memory, since the intermediate we're saving ourselves from is an entire parameter copy.

The only reasons why I can imagine this would error are 1) something wrong with the branch logic that somehow doesn't cause a problem before then, or 2) something to do with the fact that I am testing this on a nightly build of Torch.

Reply:
Strange, what version of pytorch are you using?